File: client/client-multi/client-java/src/main/java/org/kaaproject/kaa/client/persistance/KaaClientPropertiesState.java
/*
 * Copyright 2014 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.persistance;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.security.KeyPair;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.apache.avro.io.BinaryDecoder;
import org.apache.avro.io.BinaryEncoder;
import org.apache.avro.io.DecoderFactory;
import org.apache.avro.io.EncoderFactory;
import org.apache.avro.specific.SpecificDatumReader;
import org.apache.avro.specific.SpecificDatumWriter;
import org.apache.commons.codec.binary.Base64;
import org.kaaproject.kaa.client.KaaClientProperties;
import org.kaaproject.kaa.common.endpoint.gen.Topic;
import org.kaaproject.kaa.common.endpoint.gen.TopicSubscriptionInfo;
import org.kaaproject.kaa.common.endpoint.security.KeyUtil;
import org.kaaproject.kaa.common.hash.EndpointObjectHash;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class KaaClientPropertiesState implements KaaClientState {

    private static final String APP_STATE_SEQ_NUMBER = "APP_STATE_SEQ_NUMBER";
    private static final String CONFIGURATION_HASH = "CONFIGURATION_HASH";
    private static final String PROFILE_HASH = "PROFILE_HASH";

    /** The Constant logger. */
    private static final Logger LOG = LoggerFactory
            .getLogger(KaaClientPropertiesState.class);

    private static final String NF_SUBSCRIPTIONS = "nf_subscriptions";
    private static final String IS_REGISTERED = "is_registered";
    private static final String STATE_FILE_LOCATION = "state.file_location";
    private static final String CLIENT_PRIVATE_KEY_FILE_LOCATION = "keys.private";
    private static final String CLIENT_PUBLIC_KEY_FILE_LOCATION = "keys.public";

    private static final String STATE_FILE_DEFAULT = "state.properties";
    private static final String CLIENT_PRIVATE_KEY_DEFAULT = "key.private";
    private static final String CLIENT_PUBLIC_KEY_DEFAULT = "key.public";

    private final KaaClientProperties state;
    private final String stateFileLocation;
    private final String privateKeyFileLocation;
    private final String clientPublicKeyFileLocation;
    private final Map<String, TopicSubscriptionInfo> nfSubscriptions = new HashMap<String, TopicSubscriptionInfo>();

    public KaaClientPropertiesState(KaaClientProperties properties) {
        super();
        stateFileLocation = properties.containsKey(STATE_FILE_LOCATION) ?
                properties.getProperty(STATE_FILE_LOCATION) :
                STATE_FILE_DEFAULT;

        privateKeyFileLocation = properties.containsKey(CLIENT_PRIVATE_KEY_FILE_LOCATION) ?
                properties.getProperty(CLIENT_PRIVATE_KEY_FILE_LOCATION) :
                CLIENT_PRIVATE_KEY_DEFAULT;

        clientPublicKeyFileLocation = properties.containsKey(CLIENT_PUBLIC_KEY_FILE_LOCATION) ?
                properties.getProperty(CLIENT_PUBLIC_KEY_FILE_LOCATION) :
                CLIENT_PUBLIC_KEY_DEFAULT;

        state = new KaaClientProperties();
        File stateFile = new File(stateFileLocation);
        if (stateFile.exists()) {
            try {
                state.load(new FileInputStream(stateFile));

                BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(state.getProperty(NF_SUBSCRIPTIONS).getBytes(), null);
                SpecificDatumReader<TopicSubscriptionInfo> avroReader =
                        new SpecificDatumReader<TopicSubscriptionInfo>(TopicSubscriptionInfo.class);

                try {
                    TopicSubscriptionInfo decodedInfo = null;

                    //FIXME: research avro documentation for more convenient approach of iteration through encoded record
                    while (true) {
                      decodedInfo = avroReader.read(null, decoder);
                      LOG.info("Loaded {}", decodedInfo);
                      nfSubscriptions.put(decodedInfo.getTopicInfo().getId(), decodedInfo);
                    }
                } catch (Exception e) {

                }
            } catch (IOException e) {
                LOG.error("Can't load state file", e);
            }
        }
    }

    @Override
    public boolean isRegistered() {
        return Boolean.parseBoolean(state.getProperty(IS_REGISTERED,
                Boolean.FALSE.toString()));
    }

    @Override
    public void setRegistered(boolean registered) {
        state.setProperty(IS_REGISTERED, Boolean.toString(registered));
    }

    @Override
    public void persist() {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(baos, null);
        SpecificDatumWriter<TopicSubscriptionInfo> datumWriter =
                new SpecificDatumWriter<TopicSubscriptionInfo>(TopicSubscriptionInfo.class);

        try {
            for (Map.Entry<String, TopicSubscriptionInfo> cursor : nfSubscriptions.entrySet()) {
                datumWriter.write(cursor.getValue(), encoder);
                LOG.info("Persisted {}", cursor.getValue());
            }

            encoder.flush();
            state.setProperty(NF_SUBSCRIPTIONS, baos.toString());
        } catch (IOException e) {
            LOG.error("Can't persist notification subscription info", e);
        }

        try {
            new File(stateFileLocation).renameTo(new File(stateFileLocation+ "_bckp"));
            File stateFile = new File(stateFileLocation);
            OutputStream os = new FileOutputStream(stateFile);
            state.store(os, null);
            os.close();
        } catch (IOException e) {
            LOG.error("Can't persist state file", e);
        }
    }

    @Override
    public PrivateKey getPrivateKey() {
        PrivateKey privateKey = null;
        File f = new File(privateKeyFileLocation);
        if (f.exists()) {
            try {
                privateKey = KeyUtil.getPrivate(f);
            } catch (Exception e) {
                LOG.info("Error loading Client Private Key", e);
                throw new RuntimeException(e);
            }
        }
        if (privateKey == null) {
            LOG.info("Generating Client Key pair");
            KeyPair kp = KeyUtil.generateKeyPair(privateKeyFileLocation, clientPublicKeyFileLocation);
            privateKey = kp.getPrivate();
        }
        return privateKey;
    }

    @Override
    public PublicKey getPublicKey() {
        PublicKey publicKey = null;
        File f = new File(clientPublicKeyFileLocation);
        if (f.exists()) {
            try {
                publicKey = KeyUtil.getPublic(f);
            } catch (Exception e) {
                LOG.info("Error loading Client Public Key", e);
                throw new RuntimeException(e);
            }
        }
        if (publicKey == null) {
            LOG.info("Generating Client Key pair");
            KeyPair kp = KeyUtil.generateKeyPair(privateKeyFileLocation, clientPublicKeyFileLocation);
            publicKey = kp.getPublic();
        }
        return publicKey;
    }

    @Override
    public int getAppStateSeqNumber() {
        return Integer.parseInt(state.getProperty(APP_STATE_SEQ_NUMBER, "1"));
    }

    @Override
    public EndpointObjectHash getConfigurationHash() {
        return EndpointObjectHash.fromBytes(Base64.decodeBase64(state.getProperty(CONFIGURATION_HASH, Base64.encodeBase64String(new byte[0]))));
    }

    @Override
    public EndpointObjectHash getProfileHash() {
        return EndpointObjectHash.fromBytes(Base64.decodeBase64(state.getProperty(PROFILE_HASH, Base64.encodeBase64String(new byte[0]))));
    }

    @Override
    public void setAppStateSeqNumber(int appStateSeqNumber) {
        state.setProperty(APP_STATE_SEQ_NUMBER, Integer.toString(appStateSeqNumber));
    }

    @Override
    public void setConfigurationHash(EndpointObjectHash hash) {
        state.setProperty(CONFIGURATION_HASH, Base64.encodeBase64String(hash.getData()));
    }

    @Override
    public void setProfileHash(EndpointObjectHash hash) {
        state.setProperty(PROFILE_HASH, Base64.encodeBase64String(hash.getData()));
    }

    @Override
    public void addTopic(Topic topic) {
        TopicSubscriptionInfo subscriptionInfo = nfSubscriptions.get(topic.getId());

        if (subscriptionInfo == null) {
            subscriptionInfo = TopicSubscriptionInfo.newBuilder().setTopicInfo(topic).setSeqNumber(0).build();
            nfSubscriptions.put(topic.getId(), subscriptionInfo);
            LOG.info("Adding new seqNumber 0 for {} subscription", topic.getId());
        }
    }

    @Override
    public void removeTopic(String topicId) {
        if (nfSubscriptions.remove(topicId) != null) {
            LOG.info("Removed subscription info for {}", topicId);
        }
    }

    @Override
    public void updateTopicSubscriptionInfo(String topicId, Integer sequenceNumber) {
        TopicSubscriptionInfo subscriptionInfo = nfSubscriptions.get(topicId);

        if (subscriptionInfo != null) {
            subscriptionInfo.setSeqNumber(Math.max(sequenceNumber, subscriptionInfo.getSeqNumber()));
            nfSubscriptions.put(topicId, subscriptionInfo);
            LOG.info("Updated seqNumber to {} for {} subscription", subscriptionInfo.getSeqNumber(), topicId);
        }
    }

    @Override
    public Map<String, Integer> getNfSubscriptions() {
        HashMap<String, Integer> subscriptions = new HashMap<String, Integer>();

        for (Map.Entry<String, TopicSubscriptionInfo> cursor : nfSubscriptions.entrySet()) {
            subscriptions.put(cursor.getKey(), cursor.getValue().getSeqNumber());
        }

        return subscriptions;
    }

    @Override
    public List<Topic> getTopics() {
        List<Topic> topics = new LinkedList<Topic>();

        for (Map.Entry<String, TopicSubscriptionInfo> cursor : nfSubscriptions.entrySet()) {
            topics.add(cursor.getValue().getTopicInfo());
        }

        return topics;
    }
}

File: client/client-multi/client-objective-c/Kaa/persistance/KaaClientPropertiesState.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "KaaClientPropertiesState.h"
#import "KeyPair.h"
#import "EndpointAccessToken.h"
#import "EndpointKeyHash.h"
#import "AvroBytesConverter.h"
#import "EndpointGen.h"
#import "NSData+Conversion.h"
#import "UUID.h"
#import "KeyUtils.h"
#import "KaaLogging.h"
#import "TopicListHashCalculator.h"

#define TAG @"KaaClientPropertiesState >>>"

#define APP_STATE_SEQ_NUMBER    @"APP_STATE_SEQ_NUMBER"
#define PROFILE_HASH            @"PROFILE_HASH"
#define ENDPOINT_ACCESS_TOKEN   @"ENDPOINT_ACCESS_TOKEN"

#define ATTACHED_ENDPOINTS      @"attached_eps"
#define NF_SUBSCRIPTIONS        @"nf_subscriptions"
#define IS_REGISTERED           @"is_registered"
#define IS_ATTACHED             @"is_attached"
#define EVENT_SEQ_NUM           @"event.seq.num"
#define PROPERTIES_HASH         @"properties.hash"
#define TOPIC_LIST              @"topic.list"
#define TOPIC_LIST_HASH         @"topic.list.hash"
#define NEED_PROFILE_RESYNC     @"need.profile.resync"

@interface KaaClientPropertiesState ()

@property (nonatomic, strong) id<KAABase64> base64;
@property (nonatomic, strong) NSMutableDictionary *state;
@property (nonatomic, strong) NSString *stateFileLocation;
@property (nonatomic, strong) NSMutableDictionary *topicDictionary;            //<int64_t, Topic> as key-value
@property (nonatomic, strong) NSMutableDictionary *notificationSubscriptions;  //<int64_t, int32_t> as key-value
@property (nonatomic, strong) KeyPair *keyPair;
@property (nonatomic) BOOL isConfigVersionUpdated;
@property (nonatomic) BOOL hasUpdate;

- (void)setPropertiesHash:(NSData *)hash;
- (BOOL)isSDKProperyListUpdated:(KaaClientProperties *)sdkProperties;
- (void)parseNotificationSubscriptions;
- (void)parseTopics;
- (KeyPair *)getOrGenerateKeyPair;
- (void)deleteFileAtPath:(NSString *)path;

- (void)setStateStringValue:(NSString *)value forPropertyKey:(NSString *)propertyKey;
- (void)setStateBooleanValue:(BOOL)value forPropertyKey:(NSString *)propertyKey;

@end

@implementation KaaClientPropertiesState

@synthesize isRegistred = _isRegistred;
@synthesize privateKey = _privateKey;
@synthesize publicKey = _publicKey;
@synthesize endpointKeyHash = _endpointKeyHash;
@synthesize appStateSequenceNumber = _appStateSequenceNumber;
@synthesize profileHash = _profileHash;
@synthesize attachedEndpoints = _attachedEndpoints;
@synthesize endpointAccessToken = _endpointAccessToken;
@synthesize eventSequenceNumber = _eventSequenceNumber;
@synthesize isAttachedToUser = _isAttachedToUser;
@synthesize topicListHash = _topicListHash;
@synthesize needProfileResync = _needProfileResync;

- (instancetype)initWithBase64:(id<KAABase64>)base64 clientProperties:(KaaClientProperties *)properties {
    self = [super init];
    if (self) {
        self.base64 = base64;
        self.notificationSubscriptions = [NSMutableDictionary dictionary];
        self.topicDictionary = [NSMutableDictionary dictionary];
        self.attachedEndpoints = [NSMutableDictionary dictionary];
        self.isConfigVersionUpdated = NO;
        
        NSString *storage = [NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES) firstObject];
        NSString *stateFileName = [properties stringForKey:STATE_FILE_LOCATION_KEY];
        if (!stateFileName) {
            stateFileName = STATE_FILE_DEFAULT;
        }
        self.stateFileLocation = [storage stringByAppendingPathComponent:stateFileName];
        DDLogInfo(@"%@ Version: [%@], commit hash: [%@]", TAG, [properties buildVersion], [properties commitHash]);
        
        self.state = [NSMutableDictionary dictionary];
        NSFileManager *fileManager = [NSFileManager defaultManager];
        if ([fileManager fileExistsAtPath:self.stateFileLocation]) {
            //TODO: check if stateFileLocation is PLIST and it has correct path for both iOS and OS X
            @try {
                self.state = [[NSMutableDictionary alloc] initWithContentsOfFile:self.stateFileLocation];
                if ([self isSDKProperyListUpdated:properties]) {
                    DDLogInfo(@"%@ SDK properties were updated", TAG);
                    [self setIsRegistred:NO];
                    [self setPropertiesHash:[properties propertiesHash]];
                } else {
                    DDLogInfo(@"%@ SDK properties are up to date", TAG);
                }
                
                [self parseTopics];
                [self parseNotificationSubscriptions];
                
                NSString *attachedEndpointsString = self.state[ATTACHED_ENDPOINTS];
                if (attachedEndpointsString) {
                    NSArray *endpointsList = [attachedEndpointsString componentsSeparatedByString:@","];
                    for (NSString *attachedEndpoint in endpointsList) {
                        if (attachedEndpoint.length == 0) {
                            continue;
                        }
                        NSArray *splittedValues = [attachedEndpoint componentsSeparatedByString:@":"];
                        EndpointKeyHash *keyHash = [[EndpointKeyHash alloc] initWithKeyHash:splittedValues[1]];
                        EndpointAccessToken *token = [[EndpointAccessToken alloc] initWithToken:splittedValues.firstObject];
                        self.attachedEndpoints[token] = keyHash;
                    }
                }
                
                NSString *eventSeqNumStr = self.state[EVENT_SEQ_NUM];
                if (eventSeqNumStr) {
                    int eventSeqNum = [eventSeqNumStr intValue];
                    if (eventSeqNum == 0) {
                        DDLogError(@"%@ Error occurred parsing event sequence number. Can not parse %@ to int",
                                   TAG, eventSeqNumStr);
                    }
                    self.eventSequenceNumber = eventSeqNum;
                }
                
                NSString *topicListHashString = self.state[TOPIC_LIST_HASH];
                if (topicListHashString) {
                    _topicListHash = [topicListHashString intValue];
                    
                }
            }
            @catch (NSException *exception) {
                DDLogError(@"%@ Can't load state file. Error name: %@. Error reason: %@",
                           TAG, exception.name, exception.reason);
            }
        } else {
            DDLogInfo(@"%@ First SDK start!", TAG);
            
            if (![fileManager fileExistsAtPath:storage]) {
                NSError *error;
                BOOL dirCreated = [fileManager createDirectoryAtPath:storage withIntermediateDirectories:YES attributes:nil error:&error];
                if (!dirCreated) {
                    DDLogError(@"%@ Creating directory for client state: %@", TAG, error);
                }
            }
            
            [self setPropertiesHash:properties.propertiesHash];
        }
    }
    return self;
}

- (BOOL)isConfigurationVersionUpdated {
    return self.isConfigVersionUpdated;
}

- (void)setTopicListHash:(int32_t)topicListHash {
    if (_topicListHash != topicListHash) {
        _topicListHash = topicListHash;
        self.hasUpdate = YES;
    }
}

- (BOOL)isRegistred {
    NSString *value = self.state[IS_REGISTERED];
    return value ? [value boolValue] : NO;
}

- (void)setIsRegistred:(BOOL)isRegistred {
    [self setStateBooleanValue:isRegistred forPropertyKey:IS_REGISTERED];
}

- (BOOL)needProfileResync {
    NSString *value = self.state[NEED_PROFILE_RESYNC];
    return value ? [value boolValue] : NO;
}

- (void)setNeedProfileResync:(BOOL)needProfileResync {
    [self setStateBooleanValue:needProfileResync forPropertyKey:NEED_PROFILE_RESYNC];
}

- (void)persist {
    if (!self.hasUpdate) {
        DDLogVerbose(@"%@ No updates: ignoring persist call", TAG);
        return;
    }
    
    NSMutableData *encodedData = [NSMutableData data];
    @try {
        for (Topic *topic in _topicDictionary.allValues) {
            
            size_t topicSize = [topic getSize];
            char *buffer = (char *)malloc((topicSize) * sizeof(char));
            avro_writer_t writer = avro_writer_memory(buffer, topicSize);
            if (!writer) {
                DDLogError(@"%@ Unable to allocate '%li'bytes for avro writer during persisting", TAG, topicSize);
                continue;
            }
            
            [topic serialize:writer];
            [encodedData appendBytes:writer->buf length:(NSUInteger)writer->written];
            avro_writer_free(writer);
        }
        NSData *base64Encoded = [self.base64 encodeBase64:encodedData];
        NSString *base64Str = [[NSString alloc] initWithData:base64Encoded encoding:NSUTF8StringEncoding];
        
        self.state[TOPIC_LIST] = base64Str;
        DDLogInfo(@"%@ Persisted %lli topics", TAG, (int64_t)[_topicDictionary.allValues count]);
    }
    @catch (NSException *ex) {
        DDLogError(@"%@ Can't persist topic list info. Error: %@, reason: %@", TAG, ex.name, ex.reason);
    }
    
    NSData *subscriptionsData = [NSKeyedArchiver archivedDataWithRootObject:_notificationSubscriptions];
    self.state[NF_SUBSCRIPTIONS] = [self.base64 encodedString:subscriptionsData];
    
    NSMutableString *attachedEndpointsString = [NSMutableString string];
    NSArray *keys = self.attachedEndpoints.allKeys;
    for (EndpointAccessToken *key in keys) {
        EndpointKeyHash *value = self.attachedEndpoints[key];
        [attachedEndpointsString appendString:[NSString stringWithFormat:@"%@:%@,", key.token, value.keyHash]];
    }
    self.state[ATTACHED_ENDPOINTS] = attachedEndpointsString;
    self.state[EVENT_SEQ_NUM] = [NSString stringWithFormat:@"%i", self.eventSequenceNumber];
    self.state[TOPIC_LIST_HASH] = @(_topicListHash);
    
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if ([fileManager fileExistsAtPath:self.stateFileLocation]) {
        NSString *backup = [NSString stringWithFormat:@"%@_bckp", self.stateFileLocation];
        BOOL backupResult = [fileManager copyItemAtPath:self.stateFileLocation toPath:backup error:nil];
        DDLogDebug(@"%@ Backup created: %d", TAG, backupResult);
    }
    
    BOOL peristResult = [self.state writeToFile:self.stateFileLocation atomically:YES];
    DDLogDebug(@"%@ Persist finished. Result: %d", TAG, peristResult);
}

- (NSString *)refreshEndpointAccessToken {
    NSString *newAccessToken = [UUID randomUUID];
    [self setEndpointAccessToken:newAccessToken];
    return newAccessToken;
}

- (NSData *)publicKeyAsBytes {
    [self getOrGenerateKeyPair];
    return [KeyUtils getPublicKey];
}

- (SecKeyRef)publicKey {
    return [[self getOrGenerateKeyPair] getPublicKeyRef];
}

- (SecKeyRef)privateKey {
    return [[self getOrGenerateKeyPair] getPrivateKeyRef];
}

- (EndpointKeyHash *)endpointKeyHash {
    if (!_endpointKeyHash) {
        
        //to ensure key pair generated
        [self getOrGenerateKeyPair];
        
        EndpointObjectHash *publicKeyHash = [EndpointObjectHash hashWithSHA1:[KeyUtils getPublicKey]];
        NSString *base64Str = [[NSString alloc] initWithData:[self.base64 encodeBase64:publicKeyHash.data] encoding:NSUTF8StringEncoding];
        _endpointKeyHash = [[EndpointKeyHash alloc] initWithKeyHash:base64Str];
    }
    return _endpointKeyHash;
}

- (int32_t)appStateSequenceNumber {
    NSString *value = self.state[APP_STATE_SEQ_NUMBER];
    return value ? [value intValue] : 1;
}

- (EndpointObjectHash *)profileHash {
    NSString *hash = self.state[PROFILE_HASH];
    if (!hash) {
        hash = [[NSString alloc] initWithData:[self.base64 encodeBase64:[NSData data]] encoding:NSUTF8StringEncoding];
    }
    return [EndpointObjectHash hashWithBytes:[self.base64 decodeBase64:[hash dataUsingEncoding:NSUTF8StringEncoding]]];
}

- (void)setAppStateSequenceNumber:(int32_t)appStateSequenceNumber {
    [self setStateStringValue:[@(appStateSequenceNumber) stringValue] forPropertyKey:APP_STATE_SEQ_NUMBER];
}

- (void)setProfileHash:(EndpointObjectHash *)profileHash {
    NSData *base64Data = [self.base64 encodeBase64:profileHash.data];
    NSString *base64Str = [[NSString alloc] initWithData:base64Data encoding:NSUTF8StringEncoding];
    [self setStateStringValue:base64Str forPropertyKey:PROFILE_HASH];
}

- (void)addTopic:(Topic *)topic {
    NSNumber *topicId = @(topic.id);
    if (!_topicDictionary[topicId]) {
        _topicDictionary[topicId] = topic;
        if (topic.subscriptionType == SUBSCRIPTION_TYPE_MANDATORY_SUBSCRIPTION) {
            _notificationSubscriptions[topicId] = @(0);
            DDLogInfo(@"%@ Adding new seqNumber 0 for subscription with topic id: %lli", TAG, topic.id);
        }
        _hasUpdate = YES;
        DDLogInfo(@"%@ Adding new topic with id: %lli ", TAG, topic.id);
    }
}

- (void)removeTopicId:(int64_t)topicId {
    NSNumber *key = @(topicId);
    if ([_topicDictionary objectForKey:key]) {
        [_topicDictionary removeObjectForKey:key];
        [_notificationSubscriptions removeObjectForKey:key];
        DDLogInfo(@"%@ Removed topic and subscription info for topic id: %lli", TAG, topicId);
        _hasUpdate = YES;
    }
    
}

- (void)addSubscriptionForTopicWithId:(int64_t)topicId {
    NSNumber *sequenceNumber = _notificationSubscriptions[@(topicId)];
    if (!sequenceNumber) {
        _notificationSubscriptions[@(topicId)] = @(0);
        DDLogInfo(@"%@ Added new sequence number 0 for subscription with topic id: %lli", TAG, topicId);
        _hasUpdate = YES;
    }
}

- (void)removeSubscriptionForTopicWithId:(int64_t)topicId {
    NSNumber *sequenceNumber = _notificationSubscriptions[@(topicId)];
    if (sequenceNumber) {
        [_notificationSubscriptions removeObjectForKey:@(topicId)];
        DDLogInfo(@"%@ Removed subscription info for topic with id: %lli", TAG, topicId);
        _hasUpdate = YES;
    }
}

- (BOOL)updateSubscriptionInfoForTopicId:(int64_t)topicId sequence:(int32_t)sequenceNumber {
    NSNumber *seqNum = self.notificationSubscriptions[@(topicId)];
    BOOL updated = NO;
    if (seqNum && sequenceNumber > [seqNum intValue]) {
        updated = YES;
        self.notificationSubscriptions[@(topicId)] = @(sequenceNumber);
        self.hasUpdate = YES;
        DDLogDebug(@"%@ Updated seqNumber to %i for %lld subscription", TAG, sequenceNumber, topicId);
    }
    return updated;
}

- (NSDictionary *)getNotificationSubscriptions {
    return _notificationSubscriptions;
}

- (NSArray *)getTopics {
    return _topicDictionary.allValues;
}

- (void)setAttachedEndpoints:(NSMutableDictionary *)attachedEndpoints {
    [self.attachedEndpoints removeAllObjects];
    [self.attachedEndpoints addEntriesFromDictionary:attachedEndpoints];
    self.hasUpdate = YES;
}

- (void)setEndpointAccessToken:(NSString *)endpointAccessToken {
    [self setStateStringValue:endpointAccessToken forPropertyKey:ENDPOINT_ACCESS_TOKEN];
}

- (NSString *)endpointAccessToken {
    NSString *token = self.state[ENDPOINT_ACCESS_TOKEN];
    return token ?: @"";
}

- (int32_t)getAndIncrementEventSequenceNumber {
    self.hasUpdate = YES;
    return self.eventSequenceNumber++;
}

- (void)setEventSequenceNumber:(int32_t)eventSequenceNumber {
    if (eventSequenceNumber != _eventSequenceNumber) {
        _eventSequenceNumber = eventSequenceNumber;
        self.hasUpdate = YES;
    }
}

- (BOOL)isAttachedToUser {
    NSString *value = self.state[IS_ATTACHED];
    return value ? [value boolValue] : NO;
}

- (void)setIsAttachedToUser:(BOOL)isAttachedToUser {
    [self setStateBooleanValue:isAttachedToUser forPropertyKey:IS_ATTACHED];
}

- (void)clean {
    [self setIsRegistred:NO];
    [self setNeedProfileResync:NO];
    [self deleteFileAtPath:self.stateFileLocation];
    [self deleteFileAtPath:[NSString stringWithFormat:@"%@_bckp", self.stateFileLocation]];
    self.hasUpdate = YES;
}

- (void)setPropertiesHash:(NSData *)hash {
    NSData *encodedHash = [self.base64 encodeBase64:hash];
    NSString *encodedStr = [[NSString alloc] initWithData:encodedHash encoding:NSUTF8StringEncoding];
    [self setStateStringValue:encodedStr forPropertyKey:PROPERTIES_HASH];
}

- (BOOL)isSDKProperyListUpdated:(KaaClientProperties *)sdkProperties {
    NSData *hashFromSDK = [sdkProperties propertiesHash];
    NSString *stateHash = self.state[PROPERTIES_HASH];
    if (!stateHash) {
        NSData *emptyData = [self.base64 encodeBase64:[NSData data]];
        stateHash = [[NSString alloc] initWithData:emptyData encoding:NSUTF8StringEncoding];
    }
    NSData *hashFromStateFile = [self.base64 decodeBase64:[stateHash dataUsingEncoding:NSUTF8StringEncoding]];
    return ![hashFromSDK isEqualToData:hashFromStateFile];
}

- (void)parseNotificationSubscriptions {
    NSString *encodedSubscriptions = self.state[NF_SUBSCRIPTIONS];
    if (encodedSubscriptions) {
        NSData *decodedSubscriptions = [self.base64 decodeString:encodedSubscriptions];
        [_notificationSubscriptions addEntriesFromDictionary:[NSKeyedUnarchiver unarchiveObjectWithData:decodedSubscriptions]];
    } else {
        DDLogInfo(@"%@ No subscription info found in state", TAG);
    }
}

- (void)parseTopics {
    NSString *topicListStr = self.state[TOPIC_LIST];
    if (topicListStr) {
        NSData *data = [self.base64 decodeString:topicListStr];
        Topic *topic = nil;
        avro_reader_t reader = avro_reader_memory([data bytes], [data length]);
        @try {
            while (reader->read < reader->len) {
                topic = [[Topic alloc] init];
                [topic deserialize:reader];
                DDLogDebug(@"%@ Loaded topic: %@", TAG, topic);
                _topicDictionary[@(topic.id)] = topic;
            }
        }
        @catch (NSException *exception) {
            DDLogError(@"%@ Error occurred while reading information from decoder: %@", TAG, [data hexadecimalString]);
        }
        avro_reader_free(reader);
    } else {
        DDLogInfo(@"%@ No topic list found!", TAG);
    }
}

- (KeyPair *)getOrGenerateKeyPair {
    if (self.keyPair) {
        return self.keyPair;
    }
    
    SecKeyRef private = [KeyUtils getPrivateKeyRef];
    SecKeyRef public = [KeyUtils getPublicKeyRef];
    if (private != NULL && public != NULL) {
        DDLogDebug(@"%@ Found existing key pair", TAG);
        self.keyPair = [[KeyPair alloc] initWithPrivateKeyRef:private publicKeyRef:public];
        return self.keyPair;
    }
    
    DDLogDebug(@"%@ Generating new key pair", TAG);
    self.keyPair = [KeyUtils generateKeyPair];
    return self.keyPair;
}

- (void)deleteFileAtPath:(NSString *)path {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if ([fileManager removeItemAtPath:path error:NULL] == NO) {
        DDLogWarn(@"%@ Unable to remove file at path: %@", TAG, path);
    }
}

- (void)setStateStringValue:(NSString *)value forPropertyKey:(NSString *)propertyKey {
    NSString *previous = self.state[propertyKey];
    self.state[propertyKey] = value;
    self.hasUpdate |= ![value isEqualToString:previous];
}

- (void)setStateBooleanValue:(BOOL)value forPropertyKey:(NSString *)propertyKey {
    NSString *previousRawValue = self.state[propertyKey];
    BOOL previousValue = NO;
    if (previousRawValue && previousRawValue.length > 0) {
        previousValue = [previousRawValue boolValue];
    }
    self.state[propertyKey] = value ? @"Y" : @"N";
    self.hasUpdate |= value != previousValue;
}

@end
